src/libopenrave/robot.cpp

```diff
@@ -559,42 +559,81 @@ void RobotBase::RobotStateSaver::Release()
     KinBodyStateSaver::Release();
 }
 
-///\brief removes the robot from the environment temporarily while in scope
-class EnvironmentBodyRemover
-{
-public:
-    EnvironmentBodyRemover(OpenRAVE::KinBodyPtr pBody) : _pBody(pBody), _grabbedStateSaver(pBody, OpenRAVE::KinBody::Save_GrabbedBodies) {
-        if( _pBody->IsRobot() ) {
-            // If the manip comes from a connected body, the information of which manip is active is lost once the robot
-            // is removed from env. Need to save the active manip name so that we can set it back later when the robot
-            // is re-added to the env.
-            _pBodyRobot = OpenRAVE::RaveInterfaceCast<OpenRAVE::RobotBase>(_pBody);
-            if( !!_pBodyRobot->GetActiveManipulator() ) {
-                _activeManipName = _pBodyRobot->GetActiveManipulator()->GetName();
-            }
+EnvironmentBodyRemover::EnvironmentBodyRemover(KinBodyPtr pBody, int restoreOptions) : _pBody(pBody), _environmentBodyIndex(_pBody->GetEnvironmentBodyIndex()), _restoreOptions(restoreOptions) {
+    if( _pBody->IsRobot() ) {
+        // If the manip comes from a connected body, the information of which manip is active is lost once the robot
+        // is removed from env. Need to save the active manip name so that we can set it back later when the robot
+        // is re-added to the env.
+        _pBodyRobot = RaveInterfaceCast<RobotBase>(_pBody);
+        if( !!_pBodyRobot->GetActiveManipulator() ) {
+            _activeManipName = _pBodyRobot->GetActiveManipulator()->GetName();
```

Comment:
@eisoku9618 

Any problem if we use `RobotStateSaver`'s `Save_ActiveManipulator` here?
maybe you need to restore by the name, not by the pointer?